eZ/Publish/Core/Repository/Helper/DomainMapper.php

```diff
@@ -253,37 +253,40 @@ public function buildDomainFields(
         $fields = array();
         foreach ($spiFields as $spiField) {
             // We ignore fields in content not part of the content type
-            if (!isset($fieldIdentifierMap[$spiField->fieldDefinitionId])) {
+            if (!isset($fieldDefinitionsMap[$spiField->fieldDefinitionId])) {
                 continue;
             }
 
+            $fieldDefinition = $fieldDefinitionsMap[$spiField->fieldDefinitionId];
+
             if (!empty($prioritizedLanguages) && !in_array($spiField->languageCode, $prioritizedLanguages)) {
                 // If filtering is enabled we ignore fields in other languages then $prioritizedLanguages, if:
-                if ($alwaysAvailableLanguage === null) {
-                    // Ignore field if we don't have $alwaysAvailableLanguageCode fallback
+                if ($spiField->languageCode !== $alwaysAvailableLanguage) {
+                    // field is not in $alwaysAvailableLanguageCode (or if it's null)
                     continue;
                 } elseif (!empty($fieldInFilterLanguagesMap[$spiField->fieldDefinitionId])) {
-                    // Ignore field if it exists in one of the filtered languages
-                    continue;
-                } elseif ($spiField->languageCode !== $alwaysAvailableLanguage) {
-                    // Also ignore if field is not in $alwaysAvailableLanguageCode
+                    // field already exists in one of the filtered languages
                     continue;
                 }
             }
 
-            $fields[] = new Field(
+            $fields[$fieldDefinition->position][] = new Field(
                 array(
                     'id' => $spiField->id,
                     'value' => $this->fieldTypeRegistry->getFieldType($spiField->type)
                         ->fromPersistenceValue($spiField->value),
                     'languageCode' => $spiField->languageCode,
-                    'fieldDefIdentifier' => $fieldIdentifierMap[$spiField->fieldDefinitionId],
+                    'fieldDefIdentifier' => $fieldDefinition->identifier,
                     'fieldTypeIdentifier' => $spiField->type,
                 )
             );
         }
 
-        return $fields;
+        // Sort fields by content type field definition priority
+        ksort($fields, SORT_NUMERIC);
+
+        // Flatten array
+        return array_merge(...$fields);
```

Comment:
These changes (896c49b) should rather be opened against 6.7 as it solves the random failures on postgres when tests fails with fields being in wrong order, example of failure on that outside of this PR:
https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/460386740

@alongosz  / @adamwojs ok on the approach on that or would you have done it differently?

Reply:
@andrerom I aimed to solve this rather by ORDER BY, but don't have strong opinion on that (not sure if there's a difference in performance). Btw. there is a ticket for this: https://jira.ez.no/browse/EZP-29790 :)

Reply:
We don't join by ezcontentclass_attribute table, and even if we did position does not have index.

Reply:
That's a good point in favor of your solution. Could you open a PR against 6.7 with just that change? :)

Reply:
I can :)

Reply:
Rebased and pushed now that the field sorting has been merged.